### ISSUE TITLE
RuntimeError: "This event loop is already running" with nbclient==0.8.0

### ISSUE THREAD

**Issue by taras** [jupyter-server/jupyter-scheduler#435]
Related #330 but  happening with nbclient==0.8.0

## Description

When I execute a job, I get "This event loop is already running" error in the UI.

<img width="955" alt="image" src="https://github.com/jupyter-server/jupyter-scheduler/assets/74687/441bca91-ff4c-4829-8c9a-989b3b4c5281">

In the service logs, I get the following error.

```
Sep 29 11:57:51 tubuntu jupyter-notebook[857]: [I 2023-09-29 11:57:51.712 ServerApp] Connecting to kernel a687ffa6-7465-4686-b209-7c1af177a8c2.
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]: Traceback (most recent call last):
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]:   File "/home/taras/mambaforge/envs/notebook_env/lib/python3.11/site-packages/jupyter_scheduler/executors.py", line 60, in process
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]:     self.execute()
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]:   File "/home/taras/mambaforge/envs/notebook_env/lib/python3.11/site-packages/jupyter_scheduler/executors.py", line 140, in execute
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]:     ep.preprocess(nb)
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]:   File "/home/taras/mambaforge/envs/notebook_env/lib/python3.11/site-packages/nbconvert/preprocessors/execute.py", line 94, in preprocess
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]:     with self.setup_kernel():
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]:   File "/home/taras/mambaforge/envs/notebook_env/lib/python3.11/contextlib.py", line 137, in __enter__
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]:     return next(self.gen)
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]:            ^^^^^^^^^^^^^^
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]:   File "/home/taras/mambaforge/envs/notebook_env/lib/python3.11/site-packages/nbclient/client.py", line 596, in setup_kernel
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]:     self.start_new_kernel(**kwargs)
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]:   File "/home/taras/mambaforge/envs/notebook_env/lib/python3.11/site-packages/jupyter_core/utils/__init__.py", line 168, in wrapped
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]:     return loop.run_until_complete(inner)
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]:   File "/home/taras/mambaforge/envs/notebook_env/lib/python3.11/asyncio/base_events.py", line 629, in run_until_complete
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]:     self._check_running()
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]:   File "/home/taras/mambaforge/envs/notebook_env/lib/python3.11/asyncio/base_events.py", line 588, in _check_running
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]:     raise RuntimeError('This event loop is already running')
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]: RuntimeError: This event loop is already running
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]: /home/taras/mambaforge/envs/notebook_env/lib/python3.11/site-packages/jupyter_scheduler/executors.py:64: RuntimeWarning: coroutine 'NotebookClient.async_start_new_kernel' was never awaited
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]:   self.on_failure(e)
Sep 29 11:58:02 tubuntu jupyter-notebook[36064]: RuntimeWarning: Enable tracemalloc to get the object allocation traceback
Sep 29 11:58:22 tubuntu jupyter-notebook[857]: [W 2023-09-29 11:58:22.103 ServerApp] Notebook Matrix optimization/Post Sep 29/scripts/Trust_ColAdd_1.ipynb is not trusted
```

## Reproduce

I'm using mamba to install Jupyter Notebook & JupyterLab and PIP to install Jupyter scheduler.

## Expected behavior

Would love to be able to run jobs :(

## Context

* [Troubleshoot Output](https://gist.github.com/taras/11944b36913ebfa7178ef4bae08643ce#file-jupyter_troubleshooting-txt)
* [Command Line Output](https://gist.github.com/taras/11944b36913ebfa7178ef4bae08643ce#file-jupyter_lab_debug-txt)

**Comment by kinaumov** [comment]
Have the exact same issue. Have tried a variety of things already, still getting the same response.

**Comment by samuelan** [comment]
try Downgrade jupyter_core to 5.3.1

**Comment by kinaumov** [comment]
Thank you very much sir @samuelan, resolved.

**Comment by taras** [comment]
It made this problem go away for me too. Thank you 🙏

**Comment by samuelan** [comment]
I found the issue with jupyter_core==5.3.2 with a different problem that I had. I haven't figured out the exact steps to reproduce the issue without sharing internal codes. If you folks have a simplified reproduction scenario you can share here, that'd be great. I commented on one PR in jupyter_core which change the way event loop was handled in run_sync function. The author of that PR was asking my repro, and so far I am not able to give him that.

**Comment by taras** [comment]
I'm guessing that reproducing my environment would probably do it. I could probably do it with some kind of sandbox that supports python.

**Comment by gbellvelocity** [comment]
I am having a similar issue, but I am installing most packages with conda install and my jupyter_core==5.4.0 and nbclient=0.8.0 and I am running into similar errors.  I have tried downgrading to jupyter_core=5.3.1 and still get the same error.  I am running user instances via jupyterhub-singleuser.  From other posts, I have also tried to set nbclient==0.7.1 and issue still persists.  I have seen other posts about asyncio python module, but they were older so assuming this is not the issue.

**Comment by samuelan** [comment]
For me, the combination that worked was 
jupyter_core==5.3.1
nbclient==0.8.0
nbclassic==1.0.0

I have seen nbclient==0.5.13 and nbclassic==0.4.7 being not compatible/broken with jupyter_core==5.3.1. So jupyter_core==5.3.1 by itself is not a silver bullet. The PR that change the event loop was in https://github.com/jupyter/jupyter_core/pull/362, and I made a comment there about that. I have not been able to reproduce the same problem with simple Python codes on my laptop, but once I do, I will post here and over there at jupyter_core about it.

**Comment by gbellvelocity** [comment]
@samuelan Thanks for the prompt response.  I did not have any luck with the settings above.  Before I started any version modifications I was at:

jupyter_core==5.4.0
nbclassic==0.5.5
nbclient==0.8.0

I am actually running on Amazon2 kernel if that helps.  I am not using containers either, long story as to why not, but customer is always right. :)

I typically create a base environment, and then I run "conda list --explicit > user.txt" and build/create user envs from that.

**Comment by samuelan** [comment]
@gbellvelocity no worries. Glad it worked for ya. 
I still don't have simple repro yet on my laptop.

**Comment by dlqqq** [comment]
@samuelan @taras Hey folks, I just ran into this issue myself. Sorry for the lack of communication from us maintainers on this issue, we've been super busy as of late 😓. First, I want to document a few things explicitly:

1. I am able to reproduce this bug with `jupyter_core==5.4.0`.
2. Downgrading to `jupyter_core==5.3.1` as recommended by @samuelan worked for me.

I see that @samuelan went the extra mile and did some great investigatory work, tracing the bug to https://github.com/jupyter/jupyter_core/pull/362. I have a pretty solid understanding of async and Jupyter, so I think I can fix this bug directly on `jupyter_core`. If it can't be done in a timely manner, I'll cut a patch release to force a version constraint like `jupyter_core>=5,<=5.3.1` to quickly fix this bug for existing users.

Thank you all for using, contributing, and debugging Jupyter Scheduler! Love the support.

**Comment by samuelan** [comment]
Thanks a lot, @dlqqq. I tried multiple rounds of reproducing the problem with simpler setup in vain. The codes that crashed for me was in executionmanager.py, where I used a
run_sync(kernel_manager.start_kernel) call. 


the PR362 in jupyter_core does not create new event loop each time, but try to reuse existing running loop. but I could not figure out what caused the sanity check in asyncio.base_events to fail.

https://github.com/python/cpython/blob/43a6e4fa4934fcc0cbd83f7f3dc1b23a5f79f24b/Lib/asyncio/base_events.py#L581-L583

supposedly after a new event loop is created for the current process/thread, it's not supposed to be running till run_until_complete or run_forever is called. 

In any case, it sounds like you got a good handle on this. I will stay tuned to your fix. Thanks a lot, @dlqqq!

**Comment by dlqqq** [comment]
Hey folks! Good news. I found the root cause of the bug. It turns out that this bug has an oddly specific set of conditions to manifest:

- POSIX environment (i.e. Linux)
- `jupyter_core>=5.3.2` (at the time of writing)
- Python <3.12

I'm closing the issue for now as it will be fixed in the next release, which will occur in the next 24 hours.

@samuelan You might be interested in the technical details of this bug. It's a three-way interaction between `multiprocessing`, `asyncio`, and the Linux kernel 😁. It's all documented in the description of #450.

**Comment by samuelan** [comment]
@dlqqq, thank you so much! No wonder I could not reproduce it locally. So it had to be on Linux, while I was on a Mac trying to reproduce. Right ahead of Halloween, this is rather wicked bug. Really appreciate solving it for us.

**Comment by guigoruiz1** [comment]
Hello, 

I am currently having the exact same error with the conditions as @dlqqq mentioned:

```
POSIX environment (i.e. Linux)
jupyter_core>=5.3.2 (at the time of writing)
Python <3.12
```

**Comment by dlqqq** [comment]
@guigoruiz1 This is fixed in `jupyter_scheduler>=2.3.0` and `jupyter-scheduler>=1.6.0`.
